lib/charms/sdcore_nms_k8s/v0/fiveg_core_gnb.py

```diff
@@ -0,0 +1,447 @@
+    """
+    try:
+        ProviderSchema(app_data=FivegCoreGnbProviderAppData(**data))
+        return True
+    except ValidationError as e:
+        logger.error("Invalid data: %s", e)
+        return False
+
+
+class FivegCoreGnbProvides(Object):
+    """Class to be instantiated by provider of the `fiveg_core_gnb`."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Create a new instance of the FivegCoreGnbProvides class.
+
+        Args:
+            charm: Juju charm
+            relation_name (str): Relation name
+        """
+        self.relation_name = relation_name
+        self.charm = charm
+        super().__init__(charm, relation_name)
+
+    def publish_gnb_config_information(
+        self, relation_id: Optional[int], tac: int, plmns: list[PLMNConfig]
+    ) -> None:
+        """Set TAC and PLMNs in the relation data.
+
+        Args:
+            relation_id (int): Relation ID (optional).
+            tac (int): Tracking Area Code.
+            plmns (list[PLMNConfig]): Configured PLMNs.
+        """
+        if not self.charm.unit.is_leader():
+            raise RuntimeError("Unit must be leader to set application relation data.")
+        if not data_matches_provider_schema(
+            data={"tac": tac, "plmns": plmns}
+        ):
+            raise ValueError(f"Invalid gNB config: {tac}, {plmns}")
+        relation = self.model.get_relation(
+            relation_name=self.relation_name, relation_id=relation_id
+        )
+        if not relation:
+            raise RuntimeError(f"Relation {self.relation_name} not created yet.")
+        relation.data[self.charm.app].update(
+            {
+                "tac": str(tac),
+                "plmns": json.dumps([plmn.asdict() for plmn in plmns])
+            }
+        )
+
+    def _get_remote_app_relation_data(self, relation_id: Optional[int]) -> Optional[dict]:
+        """Get relation data for the remote application.
+
+        Args:
+            relation_id: Juju relation ID (optional).
+
+        Returns:
+        str: Relation data for the remote application
+            or None if the relation data is invalid.
+        """
+        relation = self.model.get_relation(
+            relation_name=self.relation_name,
+            relation_id=relation_id
+        )
+
+        if not relation:
+            logger.error("No relation: %s", self.relation_name)
+            return None
+
+        if not relation.app:
+            logger.warning("No remote application in relation: %s", self.relation_name)
+            return None
+
+        remote_app_relation_data = dict(relation.data[relation.app])
+
+        if not data_matches_requirer_schema(remote_app_relation_data):
+            logger.error("Invalid relation data: %s", remote_app_relation_data)
+            return None
+
+        return remote_app_relation_data
+
+    def get_gnb_name(self, relation_id: Optional[int]) -> Optional[str]:
```

Comment:
As above, I don't think `relation_id` is really optional. Without `relation_id`, calling this method will raise an exception as soon as we have more than one relation. 
If we decide to keep optional anyway (for any reason), it should default to `None`. Otherwise calling it will be kinda awkward (`self._fiveg_core_gnb_provider.get_gnb_name(None)`)